### ISSUE TITLE
Support for larger scale datasets

### ISSUE THREAD

**Issue by albertz** [rwth-i6/returnn#1519]
Most of our existing datasets have issues loading larger scale datasets. E.g. `HDFDataset`, `OggZipDataset` or others just take too long to load for really large data. The next problem is also the shuffling/sorting logic in those datasets. Even random shuffling is done offline for the whole dataset at once, and not on-the-fly. But when partially sorting (e.g. laplace order), it's even much slower, because that also needs to go through the whole dataset to check seq lens.

Related:

* I think this can all be solved on the dataset level, independent from the backend engine (TF, PyTorch, or whatever). However, conceptually, many relevant concepts and ideas has been discussed in this issue, but specific to TF, but we can adopt then and implement them framework-independently: https://github.com/rwth-i6/returnn/issues/292
* Another related idea is the ESPnet-style dataset shuffling/sorting: https://github.com/rwth-i6/returnn/issues/1506
* Some discussion on dataset alternatives, although those discussions were a bit off-topic from the actual reported issue on supporting arbitrary data in the `HDFDataset`: https://github.com/rwth-i6/returnn/issues/331
* Similarly, in this issue on `OggZipDataset` being too slow, there have been some discussion on dataset alternatives: https://github.com/rwth-i6/returnn/issues/1256
* The logic/code in `MultiProcDataset` could be relevant for some of the ideas here, to load things in the background.
* Dataset on-the-fly post processing (e.g. speed perturbation): https://github.com/rwth-i6/returnn/issues/1505

Some ideas:

* On-the-fly laplace: If data loading and random shuffling is still feasible, but just doing laplace sorting then makes it too slow:
  * We can make another `OnTheFlyBucketingBySeqLenDataset` or so, which acts like `MetaDataset` or `MultiProcDataset`, i.e. it gets some base dataset as an option. In that base dataset, it would use random shuffling. Then there would be some pre-defined seq length buckets, and they get filled from the data from the base dataset, and once some bucket is full, it can flush it. That's exactly like `bucket_by_sequence_length` in TF, but I think I have seen the same logic also elsewhere. This is usually to directly emit the mini batches. Maybe instead of implementing this in a separate dataset, it could also be implemented as a new way for our mini-batch creation logic. (Although the minibatch creation logic is currently framework specific in RETURNN. But it doesn't need to be that way.)
  * Similarly, there could be a `OnTheFlyLaPlaceSortingDataset` or so. This would have a cache of some predefined size to cache the next N seqs (e.g. next 10k seqs or so) and then it can sort the cached seqs with laplace ordering on-the-fly. When some seq is read via `load_seqs`, it internally would fetch also a new seq from the base dataset and put it into the cache.
* We could extend `ConcatDataset` that it lazily loads the datasets and also supports to shuffle over the datasets. (Or instead also have a separate new dataset like `RandomShuffleConcatDataset` or so.) E.g. when you anyway have N HDF files, instead of having a single `HDFDataset` loading all those files, you could have N `HDFDataset`s and put them all in `ConcatDataset`.
  * The `ConcatDataset` could also preload some of the datasets. So after shuffling the datasets, it loads 3 or so of them, and once one is finished, it will load another in the background. The datasets themselves would do laplace sorting or so. Many datasets also support the cache manager. E.g. `HDFDataset` `use_cache_manager=True`. Then if disk space is a problem, it would always on-the-fly prefetch the relevant data.
* We could also introduce a new dataset, which is better suited for large scale corpora. Maybe based on Apache Parquet or Arrow or so.

**Comment by michelwi** [comment]
Related to this is also efficient local caching.

If datasets reside on a remote file server, then random access might be slow. But if the dataset is too large, then caching the whole dataset on local storage might not be feasible.

* `OnTheFlyRandomizationDataset` might be an option that reads large chunks of the data "in order" into memory and then shuffles it. But even then this might be inefficient if for each access there is some overhead (e.g. seek)
* Each dataset might get a `cache_locally` function that would do dataset related stuff (e.g. HDFDataset would cache the hdf file, ExternSprintDataset the corpus file and all audio files) to copy the data in bulk.
* For very large scale datasets (larger than memory/storage of node) we would also need to consider cache eviction.

Then we might also need to rethink how partition_epoch works (as ideally only currently cached data is in the current partition). E.g. sub-dataset based partitions instead of sequence based partitions.

**Comment by albertz** [comment]
The solution I outlined with ConcatDataset would already done all of that. Is there a problem with it?

**Comment by michelwi** [comment]
> The solution I outlined with ConcatDataset would already solve all of that.

Mostly yes. But I am unsure if / would offer the hypothesis that for some datasets copying the underlying data in one go (e.g. copy the hdf file) would be faster than reading all sequences consecutively from the dataset while data resides on a remote file server. (e.g. because of seek operations or if files need to be opened and closed)

**Comment by albertz** [comment]
The solution I outlined with ConcatDataset would already done all of that. Is there a problem with it?

**Comment by michelwi** [comment]
Yes something like that. But what I am aiming at would be an extension to be able to use it
* outside of i6
* if available local storage is smaller than the total size of the dataset

**Comment by albertz** [comment]
It would only prefetch some of the next datasets, not all of them. So that should not be a problem that there is not enough space.

**Comment by Atticus1806** [comment]
Could this maybe even be done before that? I was thinking of an extension of the segment list or a similar file which saves e.g. the sequence length and maybe even other information. 
E.g. if features are dumped into an hdf anyways the infos are already there / processable. 
This would also make some filtering (max seq len) much more clear. Of course this would require some pre-process step (which has to be done once and can be parallelized) but I would imagine that calculating batch assignments after that based on statistics saved in a file would be really cheap. 
Thinking even further this would allow some `SplitDataInto[Batches/Epochs]Job` or similar on a conditional basis if that is ever required.

**Comment by albertz** [comment]
> Could this maybe even be done before that? I was thinking of an extension of the segment list or a similar file which saves e.g. the sequence length and maybe even other information. E.g. if features are dumped into an hdf anyways the infos are already there / processable.

`MetaDataset` already supports `seq_lens_file`.

But with `HDFDataset`, that's not really necessary, as the seq lengths are also directly stored inside.

But sorting/laplace is still somewhat slow because `sort` will call the `key=get_seq_len` function for all comparisons. But when we prepare all seq lens as a vector (like we already have in `HDFDataset`) and use `numpy.sort` or so, it would be much faster.

But anyway, using the `ConcatDataset` and the method described above, all that is not really necessary. The sorting should not really be a problem anymore. Just make an individual dataset small enough (e.g. covering 100h or so) (but also not too small because laplace sorting would only be done on that level).

Note, for the ESPnet-style dataset iteration / batching (https://github.com/rwth-i6/returnn/issues/1506), you would really need such preprocessing step.

**Comment by albertz** [comment]
> outside of i6

Why do you think cache manager can not be used outside of i6?

But in any case, I think cache manager doesn't fully solve it. It just makes copying from the remote disk to local disk somewhat more efficient. But the current way that files are cleaned up on `/tmp` (namely, I think, via [tmpreaper](https://manpages.ubuntu.com/manpages/bionic/man8/tmpreaper.8.html), i.e. when atime is older than some time?), would not clean up the old files soon enough when we need more space for further datasets.

To solve that, maybe we can have a solution similar to tmpreaper: First we use a dedicated tmp cache dir for dataset files used by RETURNN (e.g. `/var/tmp/$USER/returnn/cache/datasets/$FILEHASH` or so). Instead of using cache manager directly (if at all, can be optional), it would also move the dataset files there. And then whenever some new file is going to be copied, we would check if maybe available disk space is less then some limit (maybe 20% or so? or whatever, can be configured), then we would delete files in that dedicated tmp cache dir starting from the oldest atime, until enough disk space is free again, but maybe also with some limit, e.g. to not delete files which have been accessed in the last 10 minutes or so.

**Comment by Icemole** [comment]
Hi, maybe I can provide my fifty cents with regards to this topic. I'm currently training with a large-scale dataset (ASR, ~20k hours of audio data divided into ~2k HDF files containg ~28M sequences in total), and what I noticed was that, even when `batching` was `"random"`, the training took ~19 hours to initialize (in essence, print anything other than `start training from epoch 1`).

`"laplace:.1000"` took more than two hours before I gave up, and by its functionality as well as from the messages in here, I assume that it would have taken more than 19 hours (the time it took for `"random"`) due to the added overhead of sorting. However, I would have assumed that `"random"` should have taken relatively little, since it only takes a permutation to start training. Perhaps it wasn't only the shuffling the responsible for the big loading times, but also some other variable. Maybe I/O overhead from trying to load everything on memory? Sadly I haven't been able to debug/profile the issue so far. I can rerun a similar training and debug the part where the data is loaded and batches are formed.

PS: I've seen in the [numpy docs concerning permutations](https://numpy.org/doc/stable/reference/random/generator.html#permutations) that `permutation()` copies the data, while `shuffle()` doesn't. I'm not 100% sure that the `RandomState` code that RETURNN is using is completely equivalent to `np.random.Generator.permutation()`, which is what the docs reference, but I guess the functionality should be really similar (`O(N)` + copy). Perhaps copying so much data could be an added issue, and we could use `shuffle()` instead if we're not using the original array anymore? Just an idea that came to mind.

**Comment by albertz** [comment]
Extending a bit on the file caching: I'm thinking about the problem that there are likely multiple procs which would do this logic concurrently (due to multi GPU training, but maybe also other training jobs are running concurrently or so). So we need to make sure that this works correctly.

* On copying the files from remote disk to local cache dir:
  * Make sure that there are proper locks in place, to avoid that the same file is copied concurrently. But then we also need to make sure that if some copying crashed, that we detect that correctly. Not sure if we want to use BSD/POSIX file locks for that (e.g. read [here](https://0pointer.de/blog/projects/locking.html), [here](https://stackoverflow.com/questions/28470246/python-lockf-and-flock-behaviour)) or sth else. We also have our own `LockFile` in RETURNN which works based on creating a separate lock file using `O_EXCL`, but then has some heuristic checks to detect if an old lock file is present and outdated (because if a process dies/crashes, the lock file would not get deleted automatically). BSD/POSIX file locks don't work in all cases, specifically not necessarily with NFS (and/or not reliable), that's why we have `LockFile` in RETURNN. For the file cache, they are likely on a local disk, where we likely have BSD/POSIX file locks available. But not sure if we really want to rely on that?
* On deleting files from the local cache dir:
  * Can we rely on atime being correct? I have seen cases where the disk FS is mounted with noatime for efficiency reasons. In that case, atime would not work.
  * Even with atime, maybe want to have a way to communicate more explicitly that a file is currently used?

For the extended `ConcatDataset` (or maybe we should make a separate dataset for that, `RandomShuffleConcatDataset` or so, because it's anyway almost a complete rewrite of the dataset, probably sharing more code with `MultiProcDataset` than with `ConcatDataset`):

* Should the list of datasets be just a list (could be long, e.g. having 100-1000 entries or so) or a function which is called to dynamically get a new dataset, like `get_dataset(dataset_idx)` or so?
* We don't need to implement this file caching mechanism on the base dataset (e.g. `HDFDataset`). This can be handled by `RandomShuffleConcatDataset`. E.g. for each dataset, you would additionally specify the files to cache, and then it would automatically handle those. E.g. instead of having:
    ```python
    {"class": "HDFDataset", "files": ["/nfs/big_data.hdf"]}
    ```
    You would have sth like:
    ```
    {"class": "HDFDataset", "files": [CachedFile("/nfs/big_data.hdf")]}
    ```
    Then we can recursively go through the dataset dict (via `tree` or so) and handle those `CachedFile` instances. Also, when we finished using one instance of `HDFDataset`, we can specifically communicate that it is not used anymore. Maybe there could be just a background thread which updates atime frequently, and which would stop doing this once it is not used anymore.

**Comment by albertz** [comment]
> However, I would have assumed that `"random"` should have taken relatively little, since it only takes a permutation to start training. Perhaps it wasn't only the shuffling the responsible for the big loading times, but also some other variable. Maybe I/O overhead from trying to load everything on memory?

We shouldn't just guess here but instead just check where it is actually hanging. I would also guess that it is not the random shuffling here at all here, so let's not talk further about this, unless you have verified that this is really the problem.

But anyway, using the outlined plan with `RandomShuffleConcatDataset`, all this is not really relevant.

**Comment by albertz** [comment]
We checked that `relatime` is an option we tend to use on our local disks So just relying on atime is not good then. But the background thread which updates atime frequently, as mentioned before, this thread can simply update mtime instead (just `touch` the file), and then we can rely on mtime.

**Comment by JackTemaki** [comment]
Just as random additional information, since last interspeech I still have it in mind to just try lohtse, see this slides beginning from 67: https://livejohnshopkins-my.sharepoint.com/:p:/g/personal/mwiesne2_jh_edu/EYqRDl8cIr5BsVDxi1MOW5EBUpdqh10WFkzqixPIFM63hg?rtime=951oK0x-3Eg

It is explaining how they do large scale data loading and shuffling, and also dataset mixing with chunked tar files.

**Comment by JackTemaki** [comment]
Some example implementation code: https://github.com/lhotse-speech/lhotse/blob/master/lhotse/shar/readers/lazy.py
This does not solve anything related to caching though, it is only about file format and partial loading as well as managing multi-node / multi-worker data loading

**Comment by Icemole** [comment]
> On deleting files from the local cache dir:
> Even with atime, maybe want to have a way to communicate more explicitly that a file is currently used?

How about we integrate that into some pre/post-epoch callable?

Pre-epoch callable:
1. Store the actual files for caching purposes.
2. Store a file list with the path to the cached files.

Post-epoch callable:
1. Read the file list with the cached files, and remove them.

The caching behavior would then depend on the user, who might want to keep more/less data depending on their requirements, or do additional things in the callable. Maybe RETURNN has something like this already?

**Comment by albertz** [comment]
> How about we integrate that into some pre/post-epoch callable?

I don't think there need to be such manual work. This can all be handled by `RandomShuffleConcatDataset` automatically, in the way I outlined before, using `CachedFile`. And of course, how the caching is done exactly can be configured there.

**Comment by albertz** [comment]
I wonder a bit about sub proc usages, whether we maybe should automatically use `MultiProcDataset`. But then I realized that the problem with high memory usage due to multi-processing in the dataset (#1498) multiplies even more here with `RandomShuffleConcatDataset`.

But then I also wondered further about how we actually define sub epochs here, and how this works together with our distributed training random_seed_offset logic. The point here is that we don't want to load the whole list of all seqs into memory. So maybe it makes sense to define one sub epoch as operating on one of the sub datasets (the sub dataset internally might additionally also use partition_epoch). But should random_seed_offset then use a different sub dataset per each distributed training worker? It would increase the data variability, but on the other side, the user could simply make the sub datasets larger, which would have a similar effect. With random_seed_offset, it is also somewhat maybe weird if the sub datasets in each worker are of very different sizes, leading to always using as little data as the smallest dataset. But due to random shuffling, it would always see a different portion, so maybe that's also ok.

**Comment by michelwi** [comment]
Wasn't random_seed_offset a workaround because the dataset was too slow or it was too inefficient to load and then discard (N-1)/N of the dataset.

With the improvement(s) of multiprocessing, asynchronous prefetching, and local caching, could we maybe go back to the old method and avoid these problems?

**Comment by albertz** [comment]
> Wasn't random_seed_offset a workaround because the dataset was too slow or it was too inefficient to load and then discard (N-1)/N of the dataset.

Both. For single GPU, depending on the dataset, the dataset is usually fast enough. But N is basically the factor by how much the dataset becomes slower in multi GPU, when you discard (N-1) of the data, and that can easily become the bottleneck then.

The solution that most other frameworks use for multi GPU is sharding, i.e. also only taking a subset of the data per each worker. However, the assumption is that the data is preprocessed in a way that you don't just have a single file, but you actually have M >> N files, and you can easily split and distribute the M files across the N workers. So it's actually quite similar now to the case that we are discussing here.

So I wonder now about the details. Would this mechanism make sure that the M parts are perfectly distributed without overlap on the N workers, or would each worker just independently randomly sample from the M parts? And how to deal with the problem that the number of seqs or the dataset size for each part would not be exactly the same? This is maybe only a problem because we do epoch-based training?

> With the improvement(s) of multiprocessing, asynchronous prefetching, and local caching, could we maybe go back to the old method and avoid these problems?

My thought was: We don't need to load all data and then discard (N-1) of it. It can be distributed on the files level and then no data needs to be discarded, every worker would only load the data it uses in the first place. But the details are relevant here. If every worker would just independently sample from the M files, then we are back to random_seed_offset. The alternative is to evenly distribute the M files. But in case it is evenly distributed, we would not have used some of the data in some sub epoch, because the sub epoch would stop once one of the workers finished the dataset. In case of random_seed_offset without partition_epoch, this would not be the case - at least one worker would have seen the whole dataset. With partition_epoch, I think this is also not necessarily true anymore. But in any case, the number of seqs was very much the same before (with M=1, no matter partition_epoch or not), so the amount of skipped data was maybe not too much, while now in the new situation (M>>1), it might be more. Extreme case: Imagine M files which are widely different in size, e.g. one file where there is only one sequence. So it means this one sequence would be used much more often during training relatively to other sequences. And it's also stochastic, i.e. some sequences would maybe never be used at all.

I think we should check how this is done elsewhere in case of multi GPU training. It also matters whether a framework really operates epoch-based or just step-based.

**Comment by albertz** [comment]
After thinking a bit about it, I have an idea: Let's say you have M files and N workers. For simplicity, let's say N=1 for now, and get to the case N>1 below. What is relevant is how to define a sub epoch, i.e. partition_epoch P. I assume that P << M, at least by factor 2 or so (P * 2 < M). So for every full epoch, `RandomShuffleConcatDataset` would shuffle the M datasets, and then we need to distribute the M datasets across the P subepochs. We would do that evenly. In case P * 2 = M, every sub epoch would have 2 files. Or in general, every sub epoch would have approx M/P files.

Now, for multi-GPU training, i.e. N>1, in case of random_seed_offset, if sub epochs are of different size, we would skip some data. In my previous post, I was mostly thinking about the case M/P=1. In that case, data from a very small file would be used more often (never skipping anything) while data from larger files have higher chances to be skipped, which would give some bias in training data selection which is probably unwanted. However, the larger M/P becomes, the less this should be a problem. Even already for M/P=2, I think we don't really have this problem so much anymore. But the user should be careful with selecting the partition_epoch not too large in relation to number of files. You could always split the files further to use a certan partition_epoch.

We can improve the distribution of the data over subepochs even more, by doing a bit of preprocessing: When we know the size of each file (number of seqs, number of frames, or maybe really just file size), we can more fairly evenly distribute the data.

I was also thinking about sharding performed on the file level: If you split up the number of files further across each worker (i.e. sharding), then you should make sure that even M/(P*N) >= 2 or so. But actually, I'm not really sure if there is much advantage in doing this over random_seed_offset.

**Comment by albertz** [comment]
On the technical side, as mentioned, we need to think about how multiple files are going to be defined. Note, originally, I thought about defining a sub dataset like this:
```python
{"class": "HDFDataset", "files": [CachedFile("/nfs/big_data.hdf")]}
```
And the same for every single file. Now, when multiple files (sub datasets) would be used for a sub epoch, it could actually combine them via a `ConcatDataset` just for those files. This would be a generic solution working for all types of files.

In case of `HDFDataset`, the dataset itself however already supports the combination, so there is no need for `ConcatDataset`. And it is probably also more efficient to directly use a single `HDFDataset` on multiple files. So, this is a scenario which should also be possible.

Additionally, I'm also still thinking about how or whether `MultiProcDataset` should come into the mix, either automatically (always?) or optional. E.g. when it preloads the next sub epoch already in advance (not just copying the files for the caching logic, but also actually loading the files such that the dataset is ready to be used), I think this should be done in a subprocess, such that it does not cause slowdown in the main process (esp for PyTorch, this could have a bad effect otherwise, although maybe not if the user anyway already uses num_workers>0 for the PyTorch DataLoader?). Or maybe it should anyway use a subprocess for its datasets, but additionally the user could optionally use a `MultiProcDataset` inside there, if data loading is too slow otherwise.

Maybe the user could just define a function to get the sub dataset dict for a set of files for one sub epoch.
So, maybe it could look like this:
```python
def get_sub_epoch_dataset(files_subepoch: List[str]) -> Dict[str, Any]:
  return {
    "class": "HDFDataset",
    "files": [CachedFile(fn) for fn in files_subepoch],
  }

train = {
  "class": "RandomShuffleConcatDataset",
  "files": [
    "/nfs/big_data_1.hdf",
    ...
  ],  # M files
  "get_sub_epoch_dataset": get_sub_epoch_dataset,
  "partition_epoch": P,
}
```
I deliberately made `CachedFile` also explicit here. I think this is better to handle explicit. Also, there might be other files (e.g. vocab or so) which the user also might want to cache with the same mechanism.

Or alternatively, e.g. for `OggZipDataset`, as an example for `ConcatDataset` and `MultiProcDataset` inside:
```python
def get_sub_epoch_dataset(files_subepoch: List[str]) -> Dict[str, Any]:
  return {
    "class": "MultiProcDataset",
    "num_workers": 4,
    "buffer_size": 10,
    "dataset": {
      "class": "ConcatDataset",
      "datasets": [
        {
          "class": "OggZipDataset",
          "path": CachedFile(fn),
          ...
        }
        for fn in files_subepoch
      ],
    }  # end ConcatDataset
  }  # end MultiProcDataset
```

**Comment by albertz** [comment]
I wonder if `RandomShuffleConcatDataset` is the best name. I think `ConcatFilesDataset` is maybe better. The random shuffling is optional, depending on what ordering is configured. The default ordering would actually not shuffle.

**Comment by michelwi** [comment]
If we anyway used separate processes to asynchronously load the data, do we still need a separate (set of) process(es) for each multi-gpu worker, or could all workers just request data from a shared pool of data processes? 
At least all workers on the same physical node.

**Comment by albertz** [comment]
> If we anyway used separate processes to asynchronously load the data, do we still need a separate (set of) process(es) for each multi-gpu worker, or could all workers just request data from a shared pool of data processes?
> At least all workers on the same physical node.

With or without random_seed_offset, every multi-GPU worker would potentially (or guaranteedly) operate on a different set of files. Each subepoch dataset worker of each multi-GPU worker has its own process. I don't see how something can be shared there.

Btw, I have an initial implementation ready now (#1521) but it's mostly untested so far.

**Comment by albertz** [comment]
So `ConcatFilesDataset` is in master now and can be tried out. Please report your experience and/or whether there are any problems.

**Comment by Icemole** [comment]
One question, does the `DistributeFilesDataset` (formerly known as `ConcatFilesDataset`) support `num_workers > 1`?

**Comment by Icemole** [comment]
Also what happens when we don't use `CachedFile` in the dataset? Like when the user explicitly disables caching. We'd still be able to get the main memory improvements, so it would still make sense to use the `DistributeFilesDataset`, although we wouldn't be exploiting its full potential, is that right?

**Comment by albertz** [comment]
> One question, does the `DistributeFilesDataset` (formerly known as `ConcatFilesDataset`) support `num_workers > 1`?

What `num_workers` do you refer to? PyTorch `DataLoader`? `MultiProcDataset`? `torch.distributed`? Horovod? Sth else?

What do you mean by "support"? For distributed training, we support the random_seed_offset logic. Otherwise there is no special logic currently. But there is #1538 to implement sharding for distributed training.

**Comment by albertz** [comment]
> Also what happens when we don't use `CachedFile` in the dataset? Like when the user explicitly disables caching. We'd still be able to get the main memory improvements, so it would still make sense to use the `DistributeFilesDataset`, although we wouldn't be exploiting its full potential, is that right?

It exactly as you would expect it: It just would not use the file-caching logic then. (It's not really that the user disables caching -- the user simply does not use it then.)

For CPU memory, whether files are loaded from `/var/tmp` or from NFS should not really make any difference. (I assume at least. Maybe HDF files can be `mmap`ed in one case and not in the other? But not sure if this is even done for HDF.)

But in practice, you never ever want to access the NFS during training. Either you have some other mechanism where you copy the files to the local disk beforehand, or you use `CachedFile` for this.

**Comment by Icemole** [comment]
> What num_workers do you refer to?

`MultiProcDataset` workers.

**Comment by albertz** [comment]
> One question, does the `DistributeFilesDataset` (formerly known as `ConcatFilesDataset`) support `num_workers > 1` [in `MultiProcDataset`]?

Here it is relevant where you put the `MultiProcDataset`. See my example above with `MultiProcDataset` inside the `get_sub_epoch_dataset`. There it should be not really a problem. But note that `DistributeFilesDataset` itself already uses sub procs for the sub datasets, so `MultiProcDataset` is really only needed when you really want `num_workers > 1`. I don't think you need that in case of `HDFDataset`. Specifically not when you load the data from local disk, e.g. via `CachedFile`. Which is anyway always what you should do. It might make sense when you do some CPU-heavy processing inside the dataset, e.g. `OggZipDataset` with some `audio` `post_process` function e.g. for data augmentation e.g. by doing speed perturbation or so.

`MultiProcDataset` around the whole `DistributeFilesDataset` does not work because `DistributeFilesDataset.init_seq_order` cannot work with a given `seq_order` or `seq_list`, which is what `MultiProcDataset` uses to distribute the seqs across the workers.
